text/0000-subslice-pattern-syntax.md

```diff
@@ -0,0 +1,207 @@
+[parentheses in patterns](https://github.com/rust-lang/rust/pull/48500) (`[a, (..end)]`) or an
+explicitly written start boundary (`[a, 0 .. end]`).  
+We can also make *some* disambiguation effort and, for example, interpret `..LITERAL` as a
+range because `LITERAL` can never match a subslice. Time will show if such an effort is necessary
+or not.
+
+If/when half-bounded ranges are supported in patterns, for better future compatibility we'll need
+to reserve `..PAT` as "rest of the list" in tuples and tuple structs as well, and avoid interpreting
+it as a range pattern in those positions.
+
+Note that ambiguity with unbounded ranges as they are used in expressions (`..`) already exists in
+variant `Variant(..)` and tuple `(a, b, ..)` patterns, but it's very unlikely that the `..` syntax
+will ever be used in patterns in the range meaning because it duplicates functionality of the
+wildcard pattern `_`.
+
+#### `..PAT` vs `PAT..`
+
+Originally Rust used syntax `..PAT` for subslice patterns.  
+In 2014 the syntax was changed to `PAT..` by [RFC 202](https://github.com/rust-lang/rfcs/pull/202).
+That RFC received almost no discussion before it got merged and its motivation is no longer
+relevant because arrays now use syntax `[T; N]` instead of `[T, ..N]` used in old Rust.
+
+This RFC originally proposed to switch back to `..PAT`.
+Some reasons to switch:
+- Symmetry with expressions.  
+One of the general ideas behind patterns is that destructuring with
+patterns has the same syntax as construction with expressions, if possible.  
+In expressions we already have something with the meaning "rest of the list" - functional record
+update in struct expressions `S { field1, field2, ..remaining_fields }`.
+Right now we can use `S { field1, field1, .. }` in a pattern, but can't bind the remaining fields
+as a whole (by creating a new struct type on the fly, for example). It's not inconceivable that
+in Rust 2525 we have such ability and it's reasonable to expect it using syntax `..remaining_fields`
+symmetric to expressions. It would be good for slice patterns to be consistent with it.  
+Without speculations, even if `..remaining_fields` in struct expressions and `..subslice` in slice
+patterns are not entirely the same thing, they are similar enough to keep them symmetric already.
+- Simple disambiguation.  
+When we are parsing a slice pattern and see `..` we immediately know it's
+a subslice and can parse following tokens as a pattern (unless they are `,` or `]`, then it's just
+`..`, without an attached pattern).  
+With `PAT..` we need to consume the pattern first, but that pattern may be a... `RANGE_BEGIN..`
+range pattern, then it means that we consumed too much and need to reinterpret the parsed tokens
+somehow. It's probably possible to make this work, but it's some headache that we would like to
+avoid if possible.
+
+This RFC no longer includes the addition of `..PAT` or `PAT..`, but merely `..` as it results in
```

Comment:
This "merely `..`" isn't really true based on the grammar above... it would be true if `..` was an actual pattern.